engineapi/proxy.go

```diff
@@ -1,288 +1,126 @@
-		if err := CheckInstanceManagerProxyCompatibility(im); err != nil {
-			log.WithError(err).Warnf("Use fallback client")
+		if err := CheckInstanceManagerProxySupport(im); err != nil {
+			log.WithError(err).Warn("Use fallback client")
 			shouldFallBack = true
 		}
 	}
 
 	if shouldFallBack {
 		if fallBack == nil {
-			return nil, errors.Errorf("missing fallback client for proxy")
+			return nil, errors.Errorf("missing engine client proxy fallback client")
 		}
 
 		if obj, ok := fallBack.(*EngineBinary); ok {
 			return obj, nil
 		}
 
-		return nil, errors.Errorf("BUG: invalid fallback client for proxy: %v", fallBack)
-	}
-
-	return h.GetClient(im)
-}
-
-func (h *EngineClientProxyHandler) GetClient(im *longhorn.InstanceManager) (c EngineClientProxy, err error) {
-	if im == nil {
-		return nil, errors.Errorf("BUG: failed to get proxy client due to missing instance manager")
-	}
-
-	log := getLoggerForProxy(h.logger, im.Spec.Image)
-
-	h.lock.Lock()
-	client, exist := h.Clients[im.Spec.Image]
-	h.lock.Unlock()
-	if exist {
-		err := client.Ping()
-		if err == nil {
-			return client, nil
-		}
-
-		log.WithError(err).Debug("Resetting proxy connection")
-		client.Stop(im.Spec.Image)
-
-		h.lock.Lock()
-		delete(h.Clients, im.Spec.Image)
-		h.lock.Unlock()
+		return nil, errors.Errorf("BUG: invalid engine client proxy fallback client: %v", fallBack)
 	}
 
-	return h.NewClient(log, im)
+	return NewEngineClientProxy(im, log)
 }
 
-func (h *EngineClientProxyHandler) NewClient(log logrus.FieldLogger, im *longhorn.InstanceManager) (EngineClientProxy, error) {
-	h.lock.Lock()
-	defer h.lock.Unlock()
-
-	if existClient, exist := h.Clients[im.Spec.Image]; exist {
-		// TODO: we do not delete the map to avoid nil memory panic, so need to
-		// check the ping to find any concurrent NewClient. Fix this when
-		// refactoring to use goroutine for proxy client handling.
-		err := existClient.Ping()
-		if err == nil {
-			log.Debugf("Proxy client already created")
-			return existClient, nil
-		}
-	}
+func NewEngineClientProxy(im *longhorn.InstanceManager, logger logrus.FieldLogger) (c EngineClientProxy, err error) {
+	defer func() {
+		err = errors.Wrap(err, "failed to get engine client proxy")
+	}()
 
 	isInstanceManagerRunning := im.Status.CurrentState == longhorn.InstanceManagerStateRunning
 	if !isInstanceManagerRunning {
-		return nil, errors.Errorf("cannot create new gRPC client while instance manager in %v state", im.Status.CurrentState)
+		err = errors.Errorf("%v instance manager is in %v, not running state", im.Name, im.Status.CurrentState)
+		return nil, err
 	}
 
 	hasIP := im.Status.IP != ""
 	if !hasIP {
-		return nil, errors.Errorf("cannot create new gRPC client while instance manager status IP is empty")
+		err = errors.Errorf("%v instance manager status IP is missing", im.Name)
+		return nil, err
 	}
 
-	client := &Proxy{}
-	if err := client.Start(im, log, h.ds); err != nil {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
This conext doesn't timeout which mean that all future grpc has no timeout. E.g., [snapshot list](https://github.com/longhorn/longhorn-manager/blob/fb4e45e7e6db63cf4b15f0f755a1b37616d9e678/vendor/github.com/longhorn/longhorn-instance-manager/pkg/client/proxy_snapshot.go#L66)

I am thinking that we should create a new context timeout for each grpc call. I.e, create a new context with timeout from `c.ctx` [here](https://github.com/longhorn/longhorn-manager/blob/fb4e45e7e6db63cf4b15f0f755a1b37616d9e678/vendor/github.com/longhorn/longhorn-instance-manager/pkg/client/proxy_snapshot.go#L66) and use it. What do you think?

Reply:
Because this is a proxy, the timeout context is now just delegated to the proxy server instead, so in the proxy server Proxy::SnapshotList, there is a context timeout when calling to the downstream engine gRPC server as usual, but there is no another timeout inf ProxyClient::SnapshotList as you mentioned.

Simply said, there is still a timeout, but it is on the proxy server-side instead of the client-side. This can be improved (nice to have) in the future, but remember to have the timeout longer than the one used in the proxy server to avoid the backend operation timeout earlier than its setting. 

```golang
func (p *Proxy) SnapshotList(ctx context.Context, req *rpc.ProxyEngineRequest) (resp *rpc.EngineSnapshotListProxyResponse, err error) {
	...
	recv, err := c.ReplicaList() //<-- timeout context inside
	....

	return resp, nil
}
```

Reply:
cc @c3y1huang @shuo-wu @derekbit

Reply:
Make sense. But I think it's better to have a timeout between the proxy server and the longhorn managers as well. So that the caller threads in the longhorn managers won't be blocked.

Reply:
https://github.com/longhorn/longhorn/issues/4060